junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ForkJoinPoolHierarchicalTestExecutorService.java

```diff
@@ -0,0 +1,154 @@
+import static java.util.concurrent.ForkJoinPool.defaultForkJoinWorkerThreadFactory;
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+import static org.junit.platform.engine.support.hierarchical.Node.ExecutionMode.CONCURRENT;
+
+import java.lang.Thread.UncaughtExceptionHandler;
+import java.lang.reflect.Constructor;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ForkJoinPool.ForkJoinWorkerThreadFactory;
+import java.util.concurrent.ForkJoinWorkerThread;
+import java.util.concurrent.Future;
+import java.util.concurrent.RecursiveAction;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
+import org.apiguardian.api.API;
+import org.junit.platform.commons.logging.LoggerFactory;
+import org.junit.platform.commons.util.ExceptionUtils;
+import org.junit.platform.engine.ConfigurationParameters;
+
+/**
+ * A {@link ForkJoinPool}-based
+ * {@linkplain HierarchicalTestExecutorService executor service} that executes
+ * {@linkplain TestTask test tasks} with a configured parallelism.
+ *
+ * @see ForkJoinPool
+ * @see DefaultParallelExecutionConfigurationStrategy
+ * @since 1.3
+ */
+@API(status = EXPERIMENTAL, since = "1.3")
+public class ForkJoinPoolHierarchicalTestExecutorService implements HierarchicalTestExecutorService {
+
+	private final ForkJoinPool forkJoinPool;
+
+	/**
+	 * Create a new {@code ForkJoinPoolHierarchicalTestExecutorService} based on
+	 * the supplied {@link ConfigurationParameters}.
+	 *
+	 * @see DefaultParallelExecutionConfigurationStrategy
+	 */
+	public ForkJoinPoolHierarchicalTestExecutorService(ConfigurationParameters configurationParameters) {
+		forkJoinPool = createForkJoinPool(configurationParameters);
+		LoggerFactory.getLogger(ForkJoinPoolHierarchicalTestExecutorService.class) //
+				.config(() -> "Using ForkJoinPool with parallelism of " + forkJoinPool.getParallelism());
+	}
+
+	private ForkJoinPool createForkJoinPool(ConfigurationParameters configurationParameters) {
+		ParallelExecutionConfigurationStrategy strategy = DefaultParallelExecutionConfigurationStrategy.getStrategy(
+			configurationParameters);
+		ParallelExecutionConfiguration configuration = strategy.createConfiguration(configurationParameters);
+		try {
+			// Try to use constructor available in Java >= 9
+			Constructor<ForkJoinPool> constructor = ForkJoinPool.class.getDeclaredConstructor(Integer.TYPE,
+				ForkJoinWorkerThreadFactory.class, UncaughtExceptionHandler.class, Boolean.TYPE, Integer.TYPE,
+				Integer.TYPE, Integer.TYPE, Predicate.class, Long.TYPE, TimeUnit.class);
+			return constructor.newInstance(configuration.getParallelism(), defaultForkJoinWorkerThreadFactory, null,
+				false, configuration.getCorePoolSize(), configuration.getMaxPoolSize(),
+				configuration.getMinimumRunnable(), null, configuration.getKeepAlive(), TimeUnit.SECONDS);
+		}
+		catch (Exception e) {
```

Comment:
Wouldn't catching the all inclusive Exception here also mask actually crashes in the try part of this code, eg NPEs?

Reply:
All `ParallelExecutionConfiguration` methods return primitive types thus I don't see where a NPE should come from. Am I missing something?